ocaml/xapi/xapi_pif.ml

```diff
@@ -319,33 +319,41 @@ let assert_no_other_local_pifs ~__context ~host ~network =
       )
 
 let assert_fcoe_not_in_use ~__context ~self =
-  let interface = Db.PIF.get_device ~__context ~self in
-  let output, _ =
-    Forkhelpers.execute_command_get_output !Xapi_globs.fcoe_driver
-      ["-t"; interface]
-  in
-  let output = String.trim output in
-  debug "Scsi ids on %s are: %s" interface output ;
-  let fcoe_scsids = Str.split (Str.regexp " ") output in
-  Helpers.get_my_pbds __context
-  |> List.iter (fun (_, pbd_rec) ->
-         let sr = pbd_rec.API.pBD_SR in
-         match Db.SR.get_type ~__context ~self:sr with
-         | "lvmofcoe" -> (
-           try
-             let scsid = List.assoc "SCSIid" pbd_rec.API.pBD_device_config in
-             if List.mem scsid fcoe_scsids then
-               raise
-                 (Api_errors.Server_error
-                    ( Api_errors.pif_has_fcoe_sr_in_use
-                    , [Ref.string_of self; Ref.string_of sr]
-                    )
-                 )
-           with Not_found -> ()
+  match Sys.file_exists !Xapi_globs.fcoe_driver with
+  | false ->
+      (* Does not support FCoE from XS9, presuming not in use
+       * Upgrade plugin will block upgrade with FCoE in use *)
+      ()
```

Comment:
We can't know if the `fcoe_driver` is absent intentionally or not. The original logic would raise an exception to complain the absence of the driver. But now it would pass quietly.

Reply:
That is the design,  e.g  Xapi would permit the absence of `fcoe_driver`, in which case, it think current XS does not support FCoE.   If the driver does not exist, it is by intention. 

There is another alternative by checking `Xapi_version.product_version_text_short ()` and handle accordingly.  (only for xapi, NOT for netwrokd).   This seems NOT work for XCP, which is the reason I did not prefer.  (because we have comments `In the case of XCP, all product_* fields will be blank`).  
@minglumlu  what is the suggestion?

Reply:
The suggestion is not changing this function at all.
The caller is as the following. Is it possible that there is `fcoe` in `PIF.capabilities` while the `fcoe_driver` is absent intentionally? If not, then letting it go as before (raising exception) would be fine?
```
if Db.PIF.get_capabilities ~__context ~self |> List.mem "fcoe" then
  assert_fcoe_not_in_use ~__context ~self ;
```

Reply:
`Is it possible that there is fcoe in PIF.capabilities while the fcoe_driver is absent intentionally?` 
 No, this should NOT be possible. 

`If not, then letting it go as before (raising exception) would be fine`?
No, Xapi would try to initialise the capabilities during its startup, refer to `fcoe-nics.md`